diff --git a/mesa-20.3.5/include/c99_math.h b/Mesa/include/c99_math.h
index e906c26..e14c6d9 100644
--- a/mesa-20.3.5/include/c99_math.h
+++ b/Mesa/include/c99_math.h
@@ -44,6 +44,7 @@
 #endif 
 
 
+#if 0 /* TURBOVNC */
 #if !defined(_MSC_VER) && \
     __STDC_VERSION__ < 199901L && \
     (!defined(_XOPEN_SOURCE) || _XOPEN_SOURCE < 600) && \
@@ -114,6 +115,7 @@ exp2(double d)
 }
 
 #endif /* C99 */
+#endif /* TURBOVNC */
 
 
 /*
@@ -141,6 +143,7 @@ exp2(double d)
 #endif
 
 
+#if 0 /* TURBOVNC */
 #if defined(fpclassify)
 /* ISO C99 says that fpclassify is a macro.  Assume that any implementation
  * of fpclassify, whether it's in a C99 compiler or not, will be a macro.
@@ -183,6 +186,7 @@ fpclassify(double x)
 #else
 #error "Need to include or define an fpclassify function"
 #endif
+#endif /* TURBOVNC */
 
 
 /* Since C++11, the following functions are part of the std namespace. Their C
diff --git a/mesa-20.3.5/src/glx/glxext.c b/Mesa/src/glx/glxext.c
index 34e86ae..02ca7cf 100644
--- a/mesa-20.3.5/src/glx/glxext.c
+++ b/Mesa/src/glx/glxext.c
@@ -50,9 +50,11 @@
 #include "glxextensions.h"
 
 #include "util/debug.h"
+#if 0 /* TURBOVNC */
 #ifndef GLX_USE_APPLEGL
 #include "dri_common.h"
 #endif
+#endif /* TURBOVNC */
 
 #include <X11/Xlib-xcb.h>
 #include <xcb/xcb.h>
diff --git a/mesa-20.3.5/src/glx/glxextensions.c b/Mesa/src/glx/glxextensions.c
index 10d4faf..cbce7d9 100644
--- a/mesa-20.3.5/src/glx/glxextensions.c
+++ b/Mesa/src/glx/glxextensions.c
@@ -34,7 +34,9 @@
 #include <string.h>
 #include "glxextensions.h"
 
+#if 0 /* TURBOVNC */
 #include "util/driconf.h"
+#endif /* TURBOVNC */
 
 #define SET_BIT(m,b)   (m[ (b) / 8 ] |=  (1U << ((b) % 8)))
 #define CLR_BIT(m,b)   (m[ (b) / 8 ] &= ~(1U << ((b) % 8)))
diff --git a/mesa-20.3.5/src/util/debug.c b/Mesa/src/util/debug.c
index 89ae613..ebe9060 100644
--- a/mesa-20.3.5/src/util/debug.c
+++ b/Mesa/src/util/debug.c
@@ -24,8 +24,11 @@
 #include <errno.h>
 #include <string.h>
 #include "debug.h"
+#if 0 /* TURBOVNC */
 #include "u_string.h"
+#endif /* TURBOVNC */
 
+#if 0 /* TURBOVNC */
 uint64_t
 parse_debug_string(const char *debug,
                    const struct debug_control *control)
@@ -66,6 +69,7 @@ comma_separated_list_contains(const char *list, const char *s)
 
    return false;
 }
+#endif /* TURBOVNC */
 
 /**
  * Reads an environment variable and interprets its value as a boolean.
diff --git a/mesa-20.3.5/src/util/debug.h b/Mesa/src/util/debug.h
index bbcc197..c91ac3c 100644
--- a/mesa-20.3.5/src/util/debug.h
+++ b/Mesa/src/util/debug.h
@@ -36,11 +36,13 @@ struct debug_control {
     uint64_t     flag;
 };
 
+#if 0 /* TURBOVNC */
 uint64_t
 parse_debug_string(const char *debug,
                    const struct debug_control *control);
 bool
 comma_separated_list_contains(const char *list, const char *s);
+#endif /* TURBOVNC */
 bool
 env_var_as_boolean(const char *var_name, bool default_value);
 unsigned
