diff -urN vnc_unixsrc.orig/vncserver vnc_unixsrc/vncserver
--- vnc_unixsrc.orig/vncserver	2007-04-26 17:57:06.000000000 +0700
+++ vnc_unixsrc/vncserver	2007-05-02 12:57:45.000000000 +0700
@@ -57,11 +57,21 @@
 # $vncUserDir = "/tmp/$ENV{USER}-vnc";
 
 $defaultXStartup
-    = ("#!/bin/sh\n\n".
-       "xrdb \$HOME/.Xresources\n".
-       "xsetroot -solid grey\n".
-       "xterm -geometry 80x24+10+10 -ls -title \"\$VNCDESKTOP Desktop\" &\n".
-       "twm &\n");
+    = ("#!/bin/sh\n".
+       "\n".
+       "# Red Hat Linux VNC session startup script\n".
+       "#\n".
+       "unset SESSION_MANAGER\n".
+       "exec /etc/X11/xinit/xinitrc\n".
+       "\n".
+       "# Another example of setting up desktop environment\n".
+       "#\n".
+       "# [ -x /etc/vnc/xstartup ] && exec /etc/vnc/xstartup\n".
+       "# [ -r $HOME/.Xresources ] && xrdb $HOME/.Xresources\n".
+       "# xsetroot -solid grey\n".
+       "# vncconfig -iconic &\n".
+       "# xterm -geometry 80x24+10+10 -ls -title \"$VNCDESKTOP Desktop\" &\n".
+       "# twm &\n");
 
 $xauthorityFile = "$ENV{XAUTHORITY}";
 
