diff --git a/zlib-4b9e3f0c56ce0a354bcb11f048f870f2d0fc544e/deflate.c b/zlib/deflate.c
index 5dc6a5a..e328ca9 100644
--- a/zlib-4b9e3f0c56ce0a354bcb11f048f870f2d0fc544e/deflate.c
+++ b/zlib/deflate.c
@@ -89,7 +89,7 @@ ZLIB_INTERNAL int read_buf        OF((z_streamp strm, Bytef *buf, unsigned size)
       void match_init OF((void)); /* asm code initialization */
       uInt longest_match  OF((deflate_state *s, IPos cur_match));
 #else
-local uInt longest_match  OF((deflate_state *s, IPos cur_match));
+ZLIB_INTERNAL uInt longest_match  OF((deflate_state *s, IPos cur_match));
 #endif
 
 #ifdef DEBUG
diff --git a/zlib-4b9e3f0c56ce0a354bcb11f048f870f2d0fc544e/fill_window_sse.c b/zlib/fill_window_sse.c
index 7221944..5a13bcf 100644
--- a/zlib-4b9e3f0c56ce0a354bcb11f048f870f2d0fc544e/fill_window_sse.c
+++ b/zlib/fill_window_sse.c
@@ -12,7 +12,8 @@
 
 #ifdef HAVE_SSE2
 
-#include <immintrin.h>
+#include <xmmintrin.h>
+#include <emmintrin.h>
 
 extern int read_buf        OF((z_streamp strm, Bytef *buf, unsigned size));
 
diff --git a/zlib-4b9e3f0c56ce0a354bcb11f048f870f2d0fc544e/match.c b/zlib/match.c
index cecff58..b3bec5a 100644
--- a/zlib-4b9e3f0c56ce0a354bcb11f048f870f2d0fc544e/match.c
+++ b/zlib/match.c
@@ -24,14 +24,14 @@
  * Standard longest_match
  *
  */
-local unsigned std1_longest_match(deflate_state *z_const s, IPos cur_match)
+ZLIB_INTERNAL unsigned std1_longest_match(deflate_state *z_const s, IPos cur_match)
 {
 	z_const unsigned wmask = s->w_mask;
     z_const Posf *prev = s->prev;
 
 	unsigned chain_length;
 	IPos limit;
-	int len, best_len, nice_match;
+	uInt len, best_len, nice_match;
 	unsigned char *scan, *match, *strend, scan_end, scan_end1;
 	
 	/*
@@ -142,7 +142,7 @@ local unsigned std1_longest_match(deflate_state *z_const s, IPos cur_match)
  * UNALIGNED_OK longest_match
  *
  */
-local unsigned std2_longest_match(deflate_state *z_const s, IPos cur_match)
+ZLIB_INTERNAL unsigned std2_longest_match(deflate_state *z_const s, IPos cur_match)
 {
 	z_const unsigned wmask = s->w_mask;
     z_const Posf *prev = s->prev;
@@ -150,7 +150,7 @@ local unsigned std2_longest_match(deflate_state *z_const s, IPos cur_match)
 	unsigned short scan_start, scan_end;
 	unsigned chain_length;
 	IPos limit;
-	int len, best_len, nice_match;
+	uInt len, best_len, nice_match;
 	unsigned char *scan, *strend;
 	
 	/*
@@ -265,10 +265,10 @@ local unsigned std2_longest_match(deflate_state *z_const s, IPos cur_match)
  * FASTEST-only longest_match
  *
  */
-local unsigned fastest_longest_match(deflate_state *z_const s, IPos cur_match)
+ZLIB_INTERNAL unsigned fastest_longest_match(deflate_state *z_const s, IPos cur_match)
 {
 	unsigned char *scan, *match, *strend;
-	int len;
+	uInt len;
 
 	/*
 	 * The code is optimized for HASH_BITS >= 8 and MAX_MATCH-2 multiple 
