diff --git a/libxcb-1.14/src/xcb_out.c b/libxcb/src/xcb_out.c
index c9593e5..4d190a6 100644
--- a/libxcb-1.14/src/xcb_out.c
+++ b/libxcb/src/xcb_out.c
@@ -179,7 +179,8 @@ uint32_t xcb_get_maximum_request_length(xcb_connection_t *c)
 
 static void close_fds(int *fds, unsigned int num_fds)
 {
-    for (unsigned int index = 0; index < num_fds; index++)
+    unsigned int index;
+    for (index = 0; index < num_fds; index++)
         close(fds[index]);
 }
 
